File: api/app/services/ingest_mode.py
"""S2 D3 — 운영 모드 (fast / default / precise) → vision page cap 매핑.

master plan §6 S2 D3. 사용자 결정 (Q-S2-1, Q-S2-1b, Q-S2-1e):
- default: page_cap = settings.vision_page_cap_per_doc (ENV default 50)
- fast: page_cap = JETRAG_VISION_MODE_FAST_CAP (default 10) — 메모/짧은 문서용
- precise: page_cap = 0 (페이지 무제한, 비용 한도까지)

전역 kill switch 우선 (A 안 채택):
- ENV `JETRAG_VISION_PAGE_CAP_PER_DOC=0` 또는 음수 → mode 무관 항상 0 (무한)
- 운영자가 cost cap 만으로 제어하고 싶을 때의 회복 토글.

mode 자체는 doc-level (chunk-level 미도입). 같은 doc 의 reingest 는
이전 `flags.ingest_mode` 를 default 로 prefill (router 책임).
"""

from __future__ import annotations

import os
from typing import Literal

IngestMode = Literal["fast", "default", "precise"]

# UI/router 검증용 상수 — 새 모드 추가 시 본 set 만 갱신.
INGEST_MODES: tuple[IngestMode, ...] = ("fast", "default", "precise")

_FAST_MODE_CAP_DEFAULT = 10
_FAST_MODE_CAP_ENV = "JETRAG_VISION_MODE_FAST_CAP"


def _parse_int(env_key: str, default: int) -> int:
    raw = os.environ.get(env_key)
    if raw is None or raw.strip() == "":
        return default
    try:
        return int(raw)
    except ValueError:
        return default


def resolve_page_cap(mode: IngestMode, settings) -> int:
    """mode + settings → 최종 page_cap.

    Args:
        mode: "fast" | "default" | "precise"
        settings: app.config.Settings (vision_page_cap_per_doc 만 참조)

    Returns:
        page_cap: extract.py / incremental.py 의 page cap 검사에 그대로 전달.
        - 0 또는 음수 = 무한 (page cap 영향 0)
        - 양수 = 단일 doc 안 vision call 페이지 cap

    우선순위 (A 안):
        1) settings.vision_page_cap_per_doc <= 0 → 0 (전역 kill switch)
        2) mode == "fast" → JETRAG_VISION_MODE_FAST_CAP (default 10)
        3) mode == "default" → settings.vision_page_cap_per_doc
        4) mode == "precise" → 0 (페이지 무제한)
    """
    if settings.vision_page_cap_per_doc <= 0:
        return 0
    if mode == "fast":
        return _parse_int(_FAST_MODE_CAP_ENV, _FAST_MODE_CAP_DEFAULT)
    if mode == "default":
        return settings.vision_page_cap_per_doc
    if mode == "precise":
        return 0
    raise ValueError(f"unknown ingest mode: {mode!r}")

File: supabase/functions/_shared/ingest/ingest_mode.ts
/**
 * `api/app/services/ingest_mode.py` 포팅 — 운영 모드 → vision 페이지 cap.
 *
 * | 모드 | page cap |
 * |---|---|
 * | `fast` | `JETRAG_VISION_MODE_FAST_CAP` (기본 10) — 메모·짧은 문서용 |
 * | `default` | `JETRAG_VISION_PAGE_CAP_PER_DOC` (기본 50) |
 * | `precise` | 0 = 페이지 무제한 (비용 한도까지만) |
 *
 * ## 전역 kill switch 가 모드를 이긴다
 * `JETRAG_VISION_PAGE_CAP_PER_DOC` 가 0 이하면 모드와 무관하게 0(무한)이다.
 * 운영자가 비용 한도만으로 제어하고 싶을 때의 회복 토글이다.
 *
 * ## 모드는 문서에 붙는다
 * 원본은 `page_cap_override` 를 함수 인자로 나르지만, 그 값의 출처는 언제나
 * `documents.flags.ingest_mode` 다(업로드·URL·reingest 모두 거기에 쓴다).
 * Edge 는 단계가 큐로 나뉘어 있어 인자를 나르기보다 **필요한 곳에서 flags 를 읽어
 * 다시 계산하는** 편이 맞다 — 메시지가 재시도로 낡아도 문서의 모드가 진실이다.
 */

export const INGEST_MODES = ["fast", "default", "precise"] as const;
export type IngestMode = typeof INGEST_MODES[number];

export const DEFAULT_INGEST_MODE: IngestMode = "default";

const FAST_MODE_CAP_DEFAULT = 10;

/** 원본 `ingest_mode._parse_int` — 비숫자면 기본값. 음수는 그대로 통과시킨다. */
function parseIntEnv(raw: string | undefined, dflt: number): number {
  if (raw === undefined || raw.trim() === "") return dflt;
  const n = Number.parseInt(raw, 10);
  return Number.isFinite(n) && /^\s*[+-]?\d+\s*$/.test(raw) ? n : dflt;
}

/** `flags.ingest_mode` 처럼 신뢰할 수 없는 값 → 유효한 모드. 아니면 `null`. */
export function asIngestMode(raw: unknown): IngestMode | null {
  return typeof raw === "string" && (INGEST_MODES as readonly string[]).includes(raw)
    ? raw as IngestMode
    : null;
}

/**
 * 원본 `resolve_page_cap`.
 *
 * @param pageCapPerDoc `settings.vision_page_cap_per_doc`
 */
export function resolvePageCap(
  mode: IngestMode,
  opts: { pageCapPerDoc: number; env: Record<string, string | undefined> },
): number {
  if (opts.pageCapPerDoc <= 0) return 0; // 전역 kill switch
  if (mode === "fast") {
    return parseIntEnv(opts.env["JETRAG_VISION_MODE_FAST_CAP"], FAST_MODE_CAP_DEFAULT);
  }
  if (mode === "default") return opts.pageCapPerDoc;
  if (mode === "precise") return 0;
  throw new Error(`unknown ingest mode: ${JSON.stringify(mode)}`);
}

/** 원본 `_flags_with_ingest_mode` — 기존 flags 를 보존하고 모드만 갱신. */
export function flagsWithIngestMode(
  existing: Record<string, unknown> | null | undefined,
  mode: IngestMode,
): Record<string, unknown> {
  return { ...(existing ?? {}), ingest_mode: mode };
}

/**
 * Python 문자열 `repr` — 400 메시지가 `{raw!r}` 라 따옴표까지 같아야 한다.
 * `'` 가 있고 `"` 가 없으면 큰따옴표로 감싼다(Python 규칙).
 */
export function pyStrRepr(s: string): string {
  const useDouble = s.includes("'") && !s.includes('"');
  const quote = useDouble ? '"' : "'";
  let out = "";
  for (const ch of s) {
    if (ch === "\\") out += "\\\\";
    else if (ch === quote) out += "\\" + ch;
    else if (ch === "\n") out += "\\n";
    else if (ch === "\r") out += "\\r";
    else if (ch === "\t") out += "\\t";
    else out += ch;
  }
  return quote + out + quote;
}

/**
 * 원본 `_validate_ingest_mode` — None·빈 문자열은 default, 그 외 무효값은 400.
 *
 * 예외 대신 결과를 돌려준다. 라우트마다 오류를 내는 방식이 달라(업로드는 즉시 응답,
 * reingest 는 내부 HttpError) 던지는 쪽으로 고정하면 한쪽이 어색해진다.
 */
export function validateIngestMode(
  raw: string | null | undefined,
): { mode: IngestMode } | { detail: string } {
  if (raw === null || raw === undefined || raw === "") {
    return { mode: DEFAULT_INGEST_MODE };
  }
  const mode = asIngestMode(raw);
  if (mode === null) {
    return {
      detail: `지원되지 않는 모드입니다: ${pyStrRepr(raw)} ` +
        `(허용: ${INGEST_MODES.join(", ")})`,
    };
  }
  return { mode };
}
